stonesoup/updater/kalman.py

```diff
@@ -3,831 +3,299 @@
-                self.update_lowlevel(
-                    hypothesis.prediction.mean,
-                    hypothesis.prediction.covar,
-                    measurement_function,
-                    measurement_matrix,
-                    measurement_noise_covar,
-                    hypothesis.measurement.state_vector
-                )
-            # Augment hypothesis with measurement prediction
-            hypothesis = SingleHypothesis(hypothesis.prediction,
-                                          hypothesis.measurement,
-                                          GaussianMeasurementPrediction(
-                                              meas_pred_mean, meas_pred_covar,
-                                              hypothesis.prediction.timestamp,
-                                              cross_covar)
-                                          )
-        else:
-            posterior_mean, posterior_covar, _ = \
-                self.update_on_measurement_prediction(
-                    hypothesis.prediction.mean,
-                    hypothesis.prediction.covar,
-                    hypothesis.measurement.state_vector,
-                    hypothesis.measurement_prediction.mean,
-                    hypothesis.measurement_prediction.covar,
-                    hypothesis.measurement_prediction.cross_covar,
-                    measurement_matrix,
-                    measurement_noise_covar
-                )
-
-        return GaussianStateUpdate(posterior_mean,
-                                   posterior_covar,
+            # Get the measurement model out of the measurement if it's there.
+            # If not, use the one native to the updater (which might still be
+            # none)
+            measurement_model = hypothesis.measurement.measurement_model
+            measurement_model = self._check_measurement_model(
+                measurement_model)
+
+            # Attach the measurement prediction to the hypothesis
+            hypothesis.measurement_prediction = self.predict_measurement(
+                predicted_state, measurement_model=measurement_model, **kwargs)
+
+        # Get the predicted measurement mean, innovation covariance and
+        # measurement cross-covariance
+        pred_meas = hypothesis.measurement_prediction.state_vector
+        innov_cov = hypothesis.measurement_prediction.covar
+        m_cross_cov = hypothesis.measurement_prediction.cross_covar
+
+        # Complete the calculation of the posterior
+        kalman_gain = m_cross_cov @ np.linalg.inv(innov_cov)  # This isn't
+        # optimised
+        posterior_mean = predicted_state.state_vector + \
+            kalman_gain @ (hypothesis.measurement.state_vector - pred_meas)
+        posterior_covariance = \
+            predicted_state.covar - kalman_gain @ innov_cov @ kalman_gain.T
+
+        # TODO: check and remove this line
+        # posterior_state_covariance = (P_post + P_post.T) / 2 # !!! kludge
```

Comment:
Can this be removed?

Reply:
I wanted a second opinion (still do). 

It's a practical kludge to cover over those instances where numerical instabilities leave you with a non-positive semidefinite covariance matrix. Which isn't allowed and causes onward problems. Our options are:

1. Leave it in - if the matrix is symmetric (as it should be) it has zero effect
2. Take it out - either on principle, or because it acts to cover up instances where the matrix is misbehaving. And in general we'd want to know about that.
3. Compromise - put in a flag which will implement this symmetrisation if required (or by default)

Reply:
Went for 3. Update now has a keyword flag, <force_symmetric_covariance=False>. Set to True if you want to force the symmetry of the covariance matrix.